### ISSUE TITLE
[Impeller] App crashes on wide-gamut indexed PNG decompression in Skia with wide-gamut enabled.

### ISSUE THREAD

**Issue by nerder** [flutter/flutter#133013]
### Is there an existing issue for this?

- [X] I have searched the [existing issues](https://github.com/flutter/flutter/issues)
- [X] I have read the [guide to filing a bug](https://flutter.dev/docs/resources/bug-reports)

### Steps to reproduce

Simply running an app that contains an `AssetImage` containing a large image (~1MB) causes the app to crash in IOS.

As per my logs, this is happening starting from v3.13.0.

### Expected results

The app not to crash

### Actual results

The app crashes

### Code sample

I've reproduced the bug in this small repository: https://github.com/nerder/ios_crash_asset_image

NOTE: I've also added a compressed version of the same image which instead will work. I couldn't manage to find the exact number of bytes that an image needs to be to generate the crash.

### Logs

<details><summary>Logs</summary>

```console
(lldb) 2023-08-22 03:03:47.288767+0200 Runner[8957:3341508] [VERBOSE-2:FlutterDarwinContextMetalImpeller.mm(37)] Using the Impeller rendering backend.
Warning: Unable to create restoration in progress marker file
fopen failed for data file: errno = 2 (No such file or directory)
Errors found! Invalidating cache...
Debug service listening on ws://127.0.0.1:53633/mepV6IrsO1k=/ws
Syncing files to device iPhone de XXX..
* thread #18, name = 'io.worker.6', stop reason = EXC_BAD_ACCESS (code=1, address=0x38)
    frame #0: 0x0000000104f27f5c Flutter`SkPngCodec::applyXformRow(void*, void const*) + 52
Flutter`SkPngCodec::applyXformRow:
->  0x104f27f5c <+52>: ldp    w5, w8, [x9, #0x38]
    0x104f27f60 <+56>: sxtw   x8, w8
    0x104f27f64 <+60>: add    x0, x19, x8
    0x104f27f68 <+64>: ldr    w8, [x9, #0x48]
Target 0: (Runner) stopped.
Lost connection to device.
```

</details>

### Flutter Doctor output

<details><summary>Doctor output</summary>

```console
[✓] Flutter (Channel stable, 3.13.0, on macOS 13.3.1 22E772610a darwin-arm64, locale en-IT)
    • Flutter version 3.13.0 on channel stable at /Users/nerder/Development/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision efbf63d9c6 (6 days ago), 2023-08-15 21:05:06 -0500
    • Engine revision 1ac611c64e
    • Dart version 3.1.0
    • DevTools version 2.25.0

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.0)
    • Android SDK at /Users/nerder/Library/Android/sdk
    • Platform android-33, build-tools 33.0.0
    • Java binary at: /Users/nerder/Library/Application
      Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/223.8836.35.2231.10406996/Android
      Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E300c
    • CocoaPods version 1.12.1

[✓] Android Studio (version 2022.2)
    • Android Studio at /Users/nerder/Library/Application
      Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/222.4459.24.2221.10121639/Android
      Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] Android Studio (version 2022.3)
    • Android Studio at /Users/nerder/Library/Application
      Support/JetBrains/Toolbox/apps/AndroidStudio/ch-0/223.8836.35.2231.10406996/Android
      Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b829.9-10027231)

[✓] IntelliJ IDEA Ultimate Edition (version 2023.2)
    • IntelliJ at /Users/nerder/Applications/JetBrains Toolbox/IntelliJ IDEA Ultimate.app
    • Flutter plugin version 74.0.4
    • Dart plugin version 231.9392.3

[✓] IntelliJ IDEA Ultimate Edition (version 2023.2)
    • IntelliJ at /Users/nerder/Library/Application
      Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/232.8660.185/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] IntelliJ IDEA Ultimate Edition (version 2023.1.5)
    • IntelliJ at /Users/nerder/Library/Application
      Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/231.9392.1/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] IntelliJ IDEA Ultimate Edition (version 2023.1.4)
    • IntelliJ at /Users/nerder/Library/Application
      Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/231.9225.16/IntelliJ IDEA.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] Connected device (1 available)
    • macOS (desktop)                 • macos                     • darwin-arm64 • macOS 13.3.1 22E772610a
      darwin-arm64

[✓] Network resources
    • All expected network resources are available.

• No issues found!
```

</details>

**Comment by dam-ease** [comment]
Hi @nerder. Thanks for filing this. I'm unable to reproduce this on the latest `master` and `stable` channels. The `image_crash.png` loads without a crash. Also tried reproducing with another image of size 2.4 MB and still can't reproduce the crash.

|  file sizes | Screenshot |
|--------|--------|
| <img width="667" alt="Screenshot 2023-08-22 at 10 39 04" src="https://github.com/flutter/flutter/assets/53122008/457cf6c7-92af-40fe-9946-9fa98b9228fe"> | <img width="667" src="https://github.com/flutter/flutter/assets/53122008/3a696491-11d0-436b-b8ce-d810bd399c64"> | 

Can you provide more information on the reproduction steps and the device the crash occurs on?

<details><summary>stable, master flutter doctor -v</summary>
<p>

```
[✓] Flutter (Channel stable, 3.13.0, on macOS 13.0 22A380 darwin-arm64, locale
    en-NG)
    • Flutter version 3.13.0 on channel stable at
      /Users/damilolaalimi/sdks/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision efbf63d9c6 (6 days ago), 2023-08-15 21:05:06 -0500
    • Engine revision 1ac611c64e
    • Dart version 3.1.0
    • DevTools version 2.25.0

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/damilolaalimi/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/damilolaalimi/Library/Android/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E300c
    • CocoaPods version 1.12.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.81.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.50.0

[✓] Connected device (4 available)
    • sdk gphone64 arm64 (mobile) • emulator-5554                        •
      android-arm64  • Android 14 (API 34) (emulator)
    • iPhone 8 (mobile)           • 218697C0-911E-4F4E-9568-C387ACB184ED • ios
      • com.apple.CoreSimulator.SimRuntime.iOS-16-4 (simulator)
    • macOS (desktop)             • macos                                •
      darwin-arm64   • macOS 13.0 22A380 darwin-arm64
    • Chrome (web)                • chrome                               •
      web-javascript • Google Chrome 116.0.5845.96

[✓] Network resources
    • All expected network resources are available.

• No issues found!
``` 
```
[!] Flutter (Channel master, 3.14.0-9.0.pre.27, on macOS 13.0 22A380 darwin-arm64, locale en-NG)
    • Flutter version 3.14.0-9.0.pre.27 on channel master at /Users/damilolaalimi/fvm/versions/master
    ! Warning: `flutter` on your path resolves to /Users/damilolaalimi/sdks/flutter/bin/flutter, which is not inside your current Flutter SDK checkout at /Users/damilolaalimi/fvm/versions/master. Consider adding /Users/damilolaalimi/fvm/versions/master/bin to the front of your path.
    ! Warning: `dart` on your path resolves to /Users/damilolaalimi/sdks/flutter/bin/dart, which is not inside your current Flutter SDK checkout at /Users/damilolaalimi/fvm/versions/master. Consider adding /Users/damilolaalimi/fvm/versions/master/bin to the front of your path.
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision e085e613c8 (74 minutes ago), 2023-08-22 02:58:24 -0400
    • Engine revision e183e8a940
    • Dart version 3.2.0 (build 3.2.0-94.0.dev)
    • DevTools version 2.26.1
    • If those were intentional, you can disregard the above warnings; however it is recommended to use "git" directly to perform update checks and upgrades.

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/damilolaalimi/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/damilolaalimi/Library/Android/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E300c
    • CocoaPods version 1.12.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.81.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.50.0

[✓] Connected device (4 available)
    • sdk gphone64 arm64 (mobile) • emulator-5554                        • android-arm64  • Android 14 (API 34) (emulator)
    • iPhone 8 (mobile)           • 218697C0-911E-4F4E-9568-C387ACB184ED • ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-4 (simulator)
    • macOS (desktop)             • macos                                • darwin-arm64   • macOS 13.0 22A380 darwin-arm64
    • Chrome (web)                • chrome                               • web-javascript • Google Chrome 116.0.5845.96

[✓] Network resources
    • All expected network resources are available.

! Doctor found issues in 1 category.
``` 

</p>
</details>

**Comment by nerder** [comment]
Hey @dam-ease, I was only able to reproduce it on a real device (Iphone 11), but since this was shipped in production we have logs from a variety of devices (ie: iPhone 14)

Since this is a memory issue I think it might be unlikely to be reproduced on the simulator, but I haven't tried tbh.

**Comment by dam-ease** [comment]
Hi @nerder. Thanks for your response. I am able to reproduce this issue on the latest `master` and `stable` channels. Without impeller, this works fine and no crash occurs.

<details><summary>Code Sample</summary>
<p>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        colorScheme: ColorScheme.fromSeed(seedColor: Colors.deepPurple),
        useMaterial3: true,
      ),
      home: const MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  const MyHomePage({super.key, required this.title});

  final String title;

  @override
  State<MyHomePage> createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Container(
        decoration: BoxDecoration(
            image: DecorationImage(
          image: AssetImage(
            "assets/images/image_crash.png",
          ),
          fit: BoxFit.cover,
        )),
        child: Container(),
      ),
    );
  }
}

``` 

</p>
</details>  

<details><summary>Log</summary>
<p>

```
* thread #15, name = 'io.worker.3', stop reason = EXC_BAD_ACCESS (code=1, address=0x38)
frame #0: 0x00000001036dbf5c Flutter`SkPngCodec::applyXformRow(void*, void const*) + 52
Flutter`SkPngCodec::applyXformRow:
-> 0x1036dbf5c <+52>: ldp w5, w8, [x9, #0x38]
0x1036dbf60 <+56>: sxtw x8, w8
0x1036dbf64 <+60>: add x0, x19, x8
0x1036dbf68 <+64>: ldr w8, [x9, #0x48]
Target 0: (Runner) stopped.
Lost connection to device.
``` 

</p>
</details>

<details><summary>stable, master flutter doctor -v</summary>
<p>

```
[✓] Flutter (Channel stable, 3.13.0, on macOS 13.0 22A380 darwin-arm64, locale
    en-NG)
    • Flutter version 3.13.0 on channel stable at
      /Users/damilolaalimi/sdks/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision efbf63d9c6 (7 days ago), 2023-08-15 21:05:06 -0500
    • Engine revision 1ac611c64e
    • Dart version 3.1.0
    • DevTools version 2.25.0

[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
    • Android SDK at /Users/damilolaalimi/Library/Android/sdk
    • Platform android-34, build-tools 34.0.0
    • ANDROID_HOME = /Users/damilolaalimi/Library/Android/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jbr/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      17.0.6+0-17.0.6b802.4-9586694)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14E300c
    • CocoaPods version 1.12.1

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2022.2)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      17.0.6+0-17.0.6b802.4-9586694)

[✓] VS Code (version 1.81.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.50.0

[✓] Connected device (4 available)
    • Damilola’s iPhone (mobile) • 71c843fa5f3dec767999f36224653c0faa78660e •
      ios            • iOS 16.6 20G75
    • iPhone 8 (mobile)          • 218697C0-911E-4F4E-9568-C387ACB184ED     •
      ios            • com.apple.CoreSimulator.SimRuntime.iOS-16-4 (simulator)
    • macOS (desktop)            • macos                                    •
      darwin-arm64   • macOS 13.0 22A380 darwin-arm64
    • Chrome (web)               • chrome                                   •
      web-javascript • Google Chrome 116.0.5845.110

[✓] Network resources
    • All expected network resources are available.

• No issues found!
``` 
/cod

</p>
</details>

**Comment by hyped0001** [comment]
Experiencing the same issue here

**Comment by dam-ease** [comment]
Hi @hyped0001. The issue needs to be identified by the domain expert and the priority label follows. See https://github.com/flutter/flutter/wiki/Issue-hygiene#priorities for more information on this.
Thank you!

**Comment by stuartmorgan** [comment]
@chinmaygarde According to the repro above this is specific to impeller; is there a reason you removed the impeller label?

(Sending back to engine queue for triage, as is standard for impeller issues.)

**Comment by chinmaygarde** [comment]
The crash is in `SkPngCodec::applyXformRow`. The codecs are identical and hit before Impeller. My guess is that `pngcrush` (or whatever was used to compress the image) got rid of the problematic bits the decoder was having trouble with.

I do think this is an engine bug. Just not an Impeller one.

**Comment by romandrahan** [comment]
confirming it's reproduced for me in https://github.com/nerder/ios_crash_asset_image on the real iPhone with both Flutter 3.13.0 and 3.13.1

though fine on 3.10.6 and 3.10.7

**Comment by jason-simmons** [comment]
This started when wide gamut support was enabled by default for iOS apps using Impeller.

`ImageDecoderImpeller` looks at the PNG asset's color space, determines that it is wide gamut, and selects `kBGR_101010x_XR_SkColorType` as the output format for decoding.

But Skia's PNG codec is unable to decode the image into that format because this PNG asset uses a palette.  Specifically, `SkSwizzler::Make` can not create a swizzler for the combination of `SkEncodedInfo::kPalette_Color` input and `kBGR_101010x_XR_SkColorType` output.

For now, apps can work around this by disabling wide gamut using this entry in `ios/Runner/Info.plist`:
```
<key>FLTEnableWideGamut</key>
<false/>
```

Within the Flutter engine, we could work around this by disabling wide gamut when decoding PNG images.  I've prototyped this at https://github.com/flutter/engine/pull/45357

@gaaclarke

**Comment by gaaclarke** [comment]
> Within the Flutter engine, we could work around this by disabling wide gamut when decoding PNG images. I've prototyped this at https://github.com/flutter/engine/pull/45357

This sounds like an oversight in the Skia decoder that should be addressed.  PNG files can definitely contain wide gamut content.  So throwing them out would be throwing out the baby with the bath water.  We had to implement that pixel format in Skia for this feature so we should fix it.

**Comment by gaaclarke** [comment]
@jason-simmons we have a unit test for display p3 PNG images: https://github.com/flutter/engine/blob/494fd7fe85a30ce4dfc28058961bb71920f514cc/lib/ui/painting/image_decoder_unittests.cc#L494

I wonder if there is something special about this png?

**Comment by jason-simmons** [comment]
The `DisplayP3Logo.png` asset used by the unit test does not contain a palette (`PLTE`) chunk.  Skia considers that input to be `SkEncodedInfo::kRGBA_Color` format, which it can decode into wide gamut output.

If the engine has to disable wide gamut for these images, then ideally it would only disable it for PNG images that use palettes.  But the `SkEncodedInfo` API that reports the encoded color type is private to Skia.

**Comment by gaaclarke** [comment]
Reading a PNG file and looking for a PLTE chunk is pretty trivial.  We don't need Skia or libPNG for that.

The other interesting thing about that problem image is that its in the "LCD Display" color profile, not sRGB or DisplayP3.  It would be nice to verify that a PNG file in the DisplayP3 color profile with a `PLTE` chunk fails.  You've seen this in the debugger so you may have a better sense that that will fail.  I think it's worth contacting Skia first since it's probably a bug they'd like to fix or seeing if we can take a stab at fixing it for them.

**Comment by gaaclarke** [comment]
Since Jason will be unavailable for a minute I implemented blocking `PLTE` png files: https://github.com/flutter/engine/pull/45399

It has a unit test that we can base our fix off of.  It will make the image not crash, but it won't render correctly either.  The colors will be off.  To fix that we'll need to fix skia.  I'll just tweak it to return a decoding error instead of crashing for now.  I think that is preferable.

**Comment by gaaclarke** [comment]
The following patch to skia appears to fix the issue:
```diff
--- a/src/codec/SkSwizzler.cpp
+++ b/src/codec/SkSwizzler.cpp
@@ -943,6 +943,9 @@ std::unique_ptr<SkSwizzler> SkSwizzler::Make(const SkEncodedInfo& encodedInfo,
                         case kRGB_565_SkColorType:
                             proc = &swizzle_index_to_565;
                             break;
+                        case kBGR_101010x_XR_SkColorType:
+                            proc = &swizzle_index_to_n32;
+                            break;
                         default:
                             return nullptr;
                     }
```

I'll verify that and submit a CL next week.  I believe this is the correct operation.  I have to verify that the colorspace transformations are happening later in the process.  Decoding the indexed values to n32 should be the correct operation, then the colorspace is used to transform that to the extended range pixel format.

**Comment by gaaclarke** [comment]
CL has been posted and Skia team was notified: https://skia-review.googlesource.com/c/skia/+/751696

**Comment by danagbemava-nc** [comment]
Updated title as this is not specific to asset images. See https://github.com/flutter/flutter/issues/134063

**Comment by jseminck** [comment]
In case more example images are needed, our app crashes when it tries to load this image: https://pbs.twimg.com/profile_images/1336717992831430657/HdqvXIvG_400x400.png

**Comment by chinmaygarde** [comment]
The compression side crash with wide-gamut is here https://github.com/flutter/flutter/issues/133942. I'll add a point of discussion for this in the weekly to figure out hot-fixes and such.

**Comment by Livinglist** [comment]
Is this fix gonna be in 3.13.4?

**Comment by chinmaygarde** [comment]
After taking into account the severity, [workaround](https://github.com/flutter/flutter/issues/133013#issuecomment-1701914395), and hot-fix window, the consensus was that this will **not** be cherry-picked into 3.13.

**Comment by sarthakydv** [comment]
It was not included in 3.13.4 and 3.13.5 as well, when can we expect this to be released?

**Comment by LeoAiolia** [comment]
What are some temporary ways to solve it？

**Comment by SethuSenthil** [comment]
The update is out on stable now so flutter upgrade should be good enough :)

**Comment by LeoAiolia** [comment]
nice,   flutter 3.13.6 is working fine

**Comment by utkarsh1097** [comment]
https://github.com/flutter/flutter/issues/135245

This issue (which I believe is related?) is not yet fixed in 3.13.6.
